src/main/java/svnserver/ext/gitea/auth/GiteaUserDB.java

```diff
@@ -0,0 +1,161 @@
+/**
+ * This file is part of git-as-svn. It is subject to the license terms
+ * in the LICENSE file found in the top-level directory of this distribution
+ * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
+ * including this file, may be copied, modified, propagated, or distributed
+ * except according to the terms contained in the LICENSE file.
+ */
+package svnserver.ext.gitea.auth;
+
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.gitea.ApiClient;
+import io.gitea.ApiException;
+import io.gitea.api.UserApi;
+import io.gitea.auth.Authentication;
+import io.gitea.auth.HttpBasicAuth;
+import io.gitea.model.UserSearchList;
+import svnserver.auth.Authenticator;
+import svnserver.auth.PlainAuthenticator;
+import svnserver.auth.User;
+import svnserver.auth.UserDB;
+import svnserver.context.SharedContext;
+import svnserver.ext.gitea.config.GiteaContext;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Gitea user authentiation.
+ *
+ * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
+ * @author Andrew Thornton <zeripath@users.noreply.github.com>
+ */
+public final class GiteaUserDB implements UserDB {
+  @NotNull
+  private static final Logger log = LoggerFactory.getLogger(GiteaUserDB.class);
+  @NotNull
+  private static final String PREFIX_USER = "user-";
+  @NotNull
+  private static final ThreadLocal<ApiClient> threadLocal = new ThreadLocal<>();
```

Comment:
I really doubt this will end up well. We reuse threads for subsequent connections, so this threadlocal will have completely unrelated stuff. Did you investigate where code that uses this is called? Maybe that functionality could just be claimed unsupported for Gitea?

Reply:
OK,

An ApiClient is actually just a holder for a username and password so it's really light and doesn't waste too much memory. Collecting the ApiClient is checked on username - so another user of the thread won't be able to get the ApiClient unless it's for the same username. This is because User is final and can't be extended to keep the password around for the remainder of the request.

It's unfortunately needed because at present there is no way to ask the Gitea API what permissions a random user has and lacks a Sudo feature - I wrote a pull-request for a sudo https://github.com/go-gitea/gitea/pull/4809 and to change the collaborator api  https://github.com/go-gitea/gitea/pull/4814 but neither particularly seem to be gaining traction.

If there's a better way to keep the session around I'd be happy to use it.

Reply:
go-gitea/gitea#4809 has been merged now in to the Gitea mainline so once Gitea v1.6 or a release is made with this patch in I will release java-gitea-api v1.6 (as appropriate) and change this patch to use the sudo. (If go-gitea/gitea#4814 is pulled we could even drop the sudo.)

Reply:
Okay, I think we can merge this PR when you get rid of threadLocal. Download of all users is weird, but since it is only used for LFS (that most likely doesn't work for different reasons), we can live with it.

Reply:
Hmm, it appears that LFS works on the svn:// protocol with the default config-gitea.example and doesn't use the lookupByExternal function.